Reflection and comptime goal (#311)
rust-lang/rust-project-goals: src/2025h2/reflection-and-comptime.md

# reflection and comptime

| Metadata         |                                                                                  |
|:-----------------|----------------------------------------------------------------------------------|
| Point of contact | @oli-obk                                                                         |
| Teams            | <!-- TEAMS WITH ASKS -->                                                         |
| Task owners      | <!-- TASK OWNERS -->                                                             |
| Status           | Proposed                                                                         |

> | Tracking issue   | https://github.com/rust-lang/rust/issues/142577                                  |

Reviewer: IIUC "Tracking issue" is supposed to be the GH project goal issue where updates will be posted.

A r-l/r tracking issue instead could be:

- others have used the following

```suggestion
| Other tracking issues | [rust-lang/rust#142577] |
```

- the "See also" metadata field is also documented and supposedly used to track progress with checkboxes and so on -- but I don't know if anyone has used it / tested to see if it works 😓

| Zulip channel    | N/A (an existing stream can be re-used or new streams can be created on request) |

## Summary

Design, implement and experimentally land a reflection scheme based on `const fn` that can only be called at compile time.

## Motivation & status quo

Creating new general purpose crates (like serialization crates, log/tracing crates, game engine state inspection crates) that should work with almost all other data structures is nontrivial today. You either need to locally implement your traits for other crates, or the other crates need to depend on you and implement your traits. This often hinders rollout and will never reach everything.

> Reflection offers a way out of this dilemma, as you can write your logic for all types, by processing the type information at runtime (or even preprocess it at compile-time) without requiring trait bounds on your functions or trait impls anywhere.

Reviewer 2: I love the *sound* of this, but I don't feel like I know exactly what use cases you are trying to enable or aiming to do. Can you give a bit more detail on what kind of things you aim to do? Is the (shiny future) goal to rewrite `PartialEq` and friends as some kind of `fn` that is written reflectively instead? Would it be generic over a data structure? Is the expectation that we could 'specialize" it to the type (and get efficiency)? Are you not sure yet, but you know the building block is X? 

If you could give a high-level sketch of what these APIs might look like, that'd be great, and bonus points if you can sketch the whole set of connections you eventually imagine.

I don't want this goal to have everything designed up front or anything, I just want to understand the big pieces in your head.

Author: I gave some more details, but 

>  Are you not sure yet, but you know the building block is X?

is a very strong motivation here 😆 I have been wanting some sort of const eval that does type stuff for a long time, and consuming types is much easier than producing new types (which I want to do at some point), so why not start there

Reviewer 3: I second the question. :D  I think what's missing is some hint of how you think things will go "out of" const-eval again. Like, I can see const-eval iterate over type information, we an expose almost arbitrary APIs to const-eval if we want to -- but then the typical use of reflection at comptime is to generate code, and const-eval can only generate values. Am I missing something obvious here or how is that supposed to work? Can we feed the result of a const-eval query into the parser and build new HIR from it? I assume that falls under the "insanely hard and requires major refactors" part, but is there a low-cost short-term version of this?

Reviewer 3: > Just getting information out in the form of constants is sufficient for an MVP that would already replace most if not all need for bevy_reflect and facet

Please go on, how can that be?
Is it that one can then use such const code to build up tables that those crates have their own logic for interpreting at runtime? Hm, I guess that's plausible. I see const-alloc coming up as a feature request very quickly here ;)

### The next 6 months

* add an attribute for `const fn` that prevents them from being called from runtime code or `const fn` without the attribute
* add basic datastructures to libcore that represent common information about types and the APIs to obtain that information

### The "shiny future" we are working towards

Create basic building blocks that allow `facet`, `bevy-reflect` and `reflect` to process types without requiring derives or trait bounds.

## Design axioms

> * Prefer procedural const-eval code over associated const based designs

Reviewer 2: I think there are interesting! I'm having a bit of trouble connecting them to the write-up above, perhaps because I don't quite get what alternatives you are weighing against.

    * We picked `const fn` in general evaluation over associated const based designs that are equally expressive but are essentially a DSL

> * Ensure privacy is upheld, modulo things like `size_of` exposing whether new private fields have been added

Reviewer 4: Starting with just "everything must be `pub`" for the MVP sounds good to me, though it will be useful to keep track of since validation at deserialisation time is a killer feature of serde. Though ... if reflection can inspect (arbitrary) attributes on public fields, then maybe `#[reflect(serde::serde(serialize_with = "path"))]` could work where reflect would just provide the outer attribute but not parse anything inside (but e.g. serde could then act based on that)

> * Ensure privacy is upheld, modulo things like `size_of` exposing whether new private fields have been added

Author: If we allow accessing private fields and types, then private fields and types are not a concept that exists anymore. Any change a crate author makes to private fields would become a breaking change now

I can see a system where we allow reflection to find From and Into impls or sth so you can get a public representation of a type with private fields. But I see no world where we'd stabilize private field accesses just like that.

| Task                         | Owner(s) or team(s) | Notes |
|------------------------------|---------------------|-------|
| Discussion and moral support | ![Team][] [lang]    |       |
| Do the work                  | oli-obk             |       |

### Design language feature to solve problem

> *Some goals propose to design a feature to solve a problem. Typically the outcome from this goal is an draft or accepted RFC. If you would like to work on an experimental implementation in-tree before an RFC is accepted, you can create a [lang team experiment](https://lang-team.rust-lang.org/how_to/experiment.html), but note that a trusted contributor is required.*

| Task                 | Owner(s) or team(s)                | Notes                                                               |
|----------------------|------------------------------------|---------------------------------------------------------------------|
| Lang-team experiment | ![Team][] [lang], [libs]           | Needs libstd data structures (lang items) to make the specialization data available |

> | Author RFC           |                                    | Not at that stage in the next 6 months                              |

Reviewer 2: ```suggestion
```

> | Lang-team champion   | ![Team][] [lang]                   | TBD |

Reviewer 2: @tmandry would need to assign a champion here

> | RFC decision         |                                    | Not at that stage in the next 6 months |

Reviewer 2: ```suggestion
```

### Implement language feature

> *If there is an accepted RFC, or you are doing a [lang-team experiment](https://lang-team.rust-lang.org/how_to/experiment.html), you commonly need someone to write the code, support from the compiler to review your PRs, and possibly lang-team design meetings to review interesting design questions. Once implementation completes we recommend a call for testing blog post.*

| Task                              | Owner(s) or team(s)                | Notes |
|-----------------------------------|------------------------------------|-------|
| Implementation                    | oli-obk |       |
| Standard reviews                  | ![Team][] [compiler]               |       |
| Lang-team champion                | ![Team][] [lang]                   |       |
| Design meeting                    | ![Team][] [lang]                   |       |
| Author call for testing blog post |  | Likely will just experiment with bevy or facet, no general call for testing |

> ## Frequently asked questions

Reviewer 2: ah, some of the details I wanted are here, though I had some trouble understanding some of these questions. e.g., "why not go full zig-style comptime" in particular felt like a bit of a nonsequitor.

Reviewer 2: maybe because I don't know zig that well

| Lang-team champion   | ![Team][] [lang]                   | TBD |

### Implement language feature

> *If there is an accepted RFC, or you are doing a [lang-team experiment](https://lang-team.rust-lang.org/how_to/experiment.html), you commonly need someone to write the code, support from the compiler to review your PRs, and possibly lang-team design meetings to review interesting design questions. Once implementation completes we recommend a call for testing blog post.*

| Task                              | Owner(s) or team(s)                | Notes |
|-----------------------------------|------------------------------------|-------|
| Implementation                    | oli-obk |       |
| Standard reviews                  | ![Team][] [compiler]               |       |
| Lang-team champion                | ![Team][] [lang]                   |       |
| Design meeting                    | ![Team][] [lang]                   |       |
| Author call for testing blog post |  | Likely will just experiment with bevy or facet, no general call for testing |

## Frequently asked questions

> ### Why do you need comptime in addition to reflection?

Author: I'm not sure what exactly you are proposing, but these operations are inherently comptime-only unless we generate a global reflection static for all types. That said, you can use this scheme to have a Reflect trait that is implemented for all types and has associated consts containing values.

#### generic const exprs

in order to use uwuflection in types in generic code you need to either write infallible code with min const generics or you need to add uwuflection bounds (Lots, think typenum) , which defeats the purpose.

### Why not go full zig-style comptime?

zig's approach to comptime from a very high level is effectively

* generate AST for all source files
* pick the `main` function and start compiling it and looking for what it needs to be compiled
* if a comptime function call is found, look only for what code that needs to compile, compile it and produce the resulting code of the comptime function
* continue the main compilation, which may now invoke the generated code and start compiling that

we do not experiment with this approach at this time, because the compiler is not set up in a way to permit proc macros from accessing type information from the current crate.
While there are ongoing refactorings that go into the direction of potentially allowing more of that in the future, that future seems to be more than 5 years away at my best guess.

> * the compiler is not set up to perform codegen while type information is already available. It possibly never will, and it would be an immense amount of work to get there. I'm doing lots of refactorings that would need to be done for sth like that anyway, even if the goal is just better incremental and general compilar architecture.

Reviewer 3: > the compiler is not set up to perform codegen while type information is already available

I am confused by this statement. We have type information during codegen. Is there a negation missing here somewhere?

Reviewer 5: ```suggestion
* the compiler is not set up to add AST nodes while type information is already available. It possibly never will, and it would be an immense amount of work to get there. I'm doing lots of refactorings that would need to be done for sth like that anyway, even if the goal is just better incremental and general compilar architecture.
```
?

Author: Yea I wanna edit everything with the feedback given on monday, but this seems like a good start
